kratos/includes/model_part.h

```diff
@@ -1988,6 +1833,62 @@ class KRATOS_API(KRATOS_CORE) ModelPart final
     ///@name Private Operations
     ///@{
 
+    template<class TReturnValueType>
+    struct ReferenceGetter
+    {
+        template<class TInputValueType>
+        inline static const TReturnValueType& Get(const TInputValueType& rInputValue) {
+            if constexpr(std::is_same_v<TReturnValueType, std::remove_cv_t<TInputValueType>>) {
+                return rInputValue;
+            } else if constexpr(std::is_same_v<TReturnValueType, std::remove_cv_t<std::decay_t<decltype(*rInputValue)>>>) {
+                return *rInputValue;
+            } else if constexpr(std::is_same_v<TReturnValueType, std::remove_cv_t<std::decay_t<decltype(**rInputValue)>>>) {
+                return **rInputValue;
+            } else {
+                static_assert(!std::is_same_v<TInputValueType, TInputValueType>, "Unsupported value type.");
+                return TReturnValueType{};
+            }
+        }
+    };
+
+    template<class TContainerType>
+    struct EntityInserter
+    {
+        ModelPart* mpModelPart;
+
+        template<class TIterator>
+        void operator()(TIterator begin, TIterator end) {
+            KRATOS_TRY
+
+            ModelPart* p_root_model_part = &mpModelPart->GetRootModelPart();
+
+            block_for_each(begin, end, [&](const auto& prEntity) {
+                const auto& r_entity = ReferenceGetter<typename TContainerType::value_type>::Get(prEntity);
+                const auto& r_entities = Container<TContainerType>::GetContainer(p_root_model_part->GetMesh()); // TODO: This is only required to only trigger a find, not a sort. Once the find is fixed, then we can simplify this.
+                auto it_found = r_entities.find(r_entity.Id());
+                KRATOS_ERROR_IF_NOT(&r_entity == &*it_found)
+                    << "attempting to add a new " << Container<TContainerType>::GetEntityName() << " with Id :"
+                    << r_entity.Id() << " to root model part " << p_root_model_part->FullName()
+                    << ", unfortunately a (different) " << Container<TContainerType>::GetEntityName()
+                    << " with the same Id already exists. [ Occurred while adding " << Container<TContainerType>::GetEntityName()
+                    << " to " << mpModelPart->FullName() << " ]."
+                    << std::endl;
+            });
+
+            //now add to the root model part
+            Container<TContainerType>::GetContainer(p_root_model_part->GetMesh()).insert(begin, end);
```

Comment:
by adding always to the root all of them, a reserve is done in the pointer vector set on the root which increases its size

Reply:
On the other hand, it is not always good to call `shrink_to_fit` as well. In the new design, if a user tries to add a range to a submodel part which is a sub-set of the one of the parent model parts, then it will not reserve the parent model parts which a super set of the range being added. I think this solves this problem.

In any case, I can expose the method `shrink_to_fit`, but I think this should be done if required only.

Reply:
sorry, but then with the current implementation the size will always grow?